js/classes/entities/midground/portal.js

```diff
@@ -1,18 +1,61 @@
 class Portal {
-    constructor(blockPos, dimension) {
+    /**
+     * @param {Vector} blockPos The position of the block the portal is on
+     * @param {Number} type The type of portal (e.g. Portal.PURPLE or Portal.YELLOW)
+     */
+    constructor(blockPos, type) {
         this.pos = Vector.blockToWorldSpace(blockPos);
-        this.size = new Vector(Block.SCALED_SIZE, Block.SCALED_SIZE * 2);
-        this.dimension = dimension;
-        this.boundingBox = new BoundingBox(this.pos, this.size);
-    };
+        this.type = type;
 
-    draw() {
-        CTX.fillStyle = "#ff0000";
-        const canvasPos = Vector.worldToCanvasSpace(this.pos);
-        CTX.fillRect(canvasPos.x, canvasPos.y, this.size.x, this.size.y);
+        const spritesheetY = this.type * Portal.SIZE.y;
+        this.animator = new Animator(Portal.SPRITESHEET, new Vector(0, spritesheetY), Portal.SIZE, 2, 0.5, true);
+
+        switch(this.type) {
+            case Portal.PURPLE:
+                this.particleEffect = new ParticleEffect(this.pos, ParticleEffect.MAGIC_PURPLE);
+                break;
+            case Portal.YELLOW:
+                this.particleEffect = new ParticleEffect(this.pos, ParticleEffect.MAGIC_YELLOW);
+                break;
```

Comment:
Switch seems a bit overkill for this. I'd prefer something like, 

```js
// Near top of file
const particleEffectMap = {
    [Portal.PURPLE]: ParticleEffect.MAGIC_PURPLE,
    [Portal.YELLOW]: ParticleEffect.MAGIC_YELLOW,
};


this.particleEffect = new ParticleEffect(this.pos, particleEffectMap[this.type]);
```

Reply:
Great idea! I'll be sure to implement that.